Zend/zend_compile.c

```diff
@@ -5316,12 +5349,72 @@ void zend_compile_declare(zend_ast *ast) /* {{{ */
 }
 /* }}} */
 
+static zend_ast **_ast_find(zend_ast **begin, zend_ast **end,
+                            zend_bool (*pred)(zend_ast *)) {
+	for (; begin < end; ++begin)
+		if (pred(*begin))
+			return begin;
+	return begin;
+}
+
+static zend_bool _is_type_decl(zend_ast *ast) {
+	return ast && ast->kind == ZEND_AST_CLASS;
+}
+
+static zend_bool _is_not_decl_stmt(zend_ast *ast) {
+	if (ast) {
+		/* todo: what else should be considered a decl stmt? */
+		switch (ast->kind) {
+		case ZEND_AST_CLASS:
+		case ZEND_AST_CONST_DECL:
+		case ZEND_AST_FUNC_DECL:
+			return 0;
+
+		default:
+			return 1;
+		}
+	}
+
+	/* todo: why are these sometimes null? */
+	return 0;
+}
+
 void zend_compile_stmt_list(zend_ast *ast) /* {{{ */
 {
-	zend_ast_list *list = zend_ast_get_list(ast);
-	uint32_t i;
-	for (i = 0; i < list->children; ++i) {
-		zend_compile_stmt(list->child[i]);
+	zend_ast_list *const list = zend_ast_get_list(ast);
+	zend_ast ** const begin = list->child;
+	zend_ast ** const end = begin + list->children;
+	zend_ast **first_decl = _ast_find(begin, end, &_is_type_decl);
+	zend_ast **p;
+
+	/* Compile opcodes before first type decl */
+	for (p = begin; p < first_decl; ++p) {
+		zend_compile_stmt(*p);
+	}
+
+	/* Compile decl stmts */
+	while (first_decl != end) {
+		zend_ast **last_decl = _ast_find(first_decl, end, &_is_not_decl_stmt);
+		HashTable unverified_types;
+		HashTable *prev_unverified_types;
+		_backup_unverified_variance_types(&unverified_types, &prev_unverified_types);
+
+		for (p = first_decl; p < last_decl; ++p) {
+			zend_compile_stmt(*p);
+		}
+
+		_compile_verify_variance(&unverified_types);
+
+		zend_hash_destroy(&unverified_types);
+		CG(unverified_types) = prev_unverified_types;
+
+		/* There can be non-consecutive type decls, so continue searching */
+		first_decl = _ast_find(last_decl, end, &_is_type_decl);
+
+		/* Compile any stmts between the two type decls (or the end) */
+		for (p = last_decl; p < first_decl; ++p) {
+			zend_compile_stmt(*p);
+		}
```

Comment:
Should this really be here? Our early binding only applies to top-level class declarations, not conditional ones, so I would kind of expect that this also holds for variance checks. On the other hand, I don't see a pressing technical reason why we can't do this.

Reply:
Maybe there is a better way, but this fixed the bug you posted, the one with multiple namespaces without using braces. We need to support variance, even if it's not top-level. Here's another example:

```php
<?php
interface A {
  function m(X $z);
}

if (true) {
  class B implements A {
    function m(Y $z) {} // ?
  }
}
```
We have a test case for things like this, but I forgot which one it was off the top of my head.

Reply:
While we do need to support variance checks for it (we certainly need that for everything), what (I think, maybe) we don't need it support is the ability to verify variance for multiple classes at once (with cyclic dependencies). I don't think we can reasonable check variance for both classes together in the example you provided, we'd have to check first A and then B, independently.

Reply:
I verified that the variance opcodes are delayed correctly for this code:

```php
<?php

interface A {
  function m(X $z): A;
}

if (true) {
  class B implements A {
    function m(X $z): B {
      return new self();
    }
  }
  class C extends B {
    function m(X $z): C {
      return new self();
    }
  }
}
```

Snippet of opcodes:
```asm
 L7    #0     JMPZ                    true                 J5                                       
 L8    #1     DECLARE_CLASS           "b"                                                           
 L13   #2     DECLARE_INHERITED_CLASS "c"                  "B"                                      
 L13   #3     VERIFY_VARIANCE         "b"                                                           
 L13   #4     VERIFY_VARIANCE         "c"                                                           
 L21   #5     RETURN<-1>              1 
```

Maybe I misunderstood your point, but it seems to work.